magento/magento2#28569:Multi-store: Missing store codes in relation t…

gallyamov opened the pull request:
### Description (*)

This is for #28569

### Related Pull Requests

N/A

### Fixed Issues (if relevant)

1. Fixes magento/magento2#28569

### Manual testing scenarios (*)

Check fixed issue for testing steps and expected results.

### Questions or comments

N/A

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

gallyamov committed:
magento/magento2#28569:Multi-store: Missing store codes in relation to a group and website
- Added stores for storeConfig query

gallyamov commented:
@magento run all tests

gallyamov commented:
@magento run Functional Tests B2B

gallyamov commented:
@magento run Functional Tests B2B

swnsma commented on app/code/Magento/StoreGraphQl/Model/Resolver/Store/StoreConfigDataProvider.php:
```diff
@@ -55,24 +56,26 @@ public function __construct(
      * @return array
      */
-    private function getBaseConfigData(StoreInterface $store) : array
+    private function prepareStoreConfigData($storeConfig): array
```
Can we add `StoreConfigInterface` to type hints in method arguments?

gallyamov replied:
Hello @swnsma , the change has been done.

swnsma commented on app/code/Magento/StoreGraphQl/Model/Resolver/Store/StoreConfigDataProvider.php:
```diff
@@ -55,24 +56,26 @@ public function __construct(
-        );
-        return $storeConfigData;
+        $stores = $this->storeConfigManager->getStoreConfigs();
+        $defaultStoreConfig = current($this->storeConfigManager->getStoreConfigs([$store->getCode()]));
```
`\Magento\Store\Model\Service\StoreConfigManager` does not cache result of `getStoreConfigs`.
Maybe we can return a list of configs and work with the list to avoid extra queries to DB.

gallyamov replied:
Hello @swnsma , the change has been done. Please let me know if any questions.

swnsma commented on app/code/Magento/StoreGraphQl/etc/schema.graphqls:
```diff
@@ -31,4 +31,5 @@ type StoreConfig @doc(description: "The type contains information about a store
     secure_base_static_url : String @doc(description: "Secure base static URL for the store")
     secure_base_media_url : String @doc(description: "Secure base media URL for the store")
     store_name : String @doc(description: "Name of the store")
+    stores: [StoreConfig] @doc(description: "Contains information about all stores")
```
Schema provides possibility to create a recursion with field 'stores' - but value will never be populated.
Do we really want to leave it as it is?

E.g. Query:
```
{
  storeConfig {
    id
    code
    website_id
    locale
    base_currency_code
    default_display_currency_code
    timezone
    weight_unit
    base_url
    base_link_url
    base_static_url
    base_media_url
    secure_base_url
    secure_base_link_url
    secure_base_static_url
    secure_base_media_url
    store_name
    stores {
      id
      code
      website_id
      locale
      base_currency_code
      default_display_currency_code
      timezone
      weight_unit
      base_url
      base_link_url
      base_static_url
      base_media_url
      secure_base_url
      secure_base_link_url
      secure_base_static_url
      secure_base_media_url
      store_name
      stores {
        id
        code
        website_id
        locale
        base_currency_code
        default_display_currency_code
        timezone
        weight_unit
        base_url
        base_link_url
        base_static_url
        base_media_url
        secure_base_url
        secure_base_link_url
        secure_base_static_url
        secure_base_media_url
        store_name
        stores {
          id
        }
      }
    }
  }
}
```

Response:
```
{
  "data": {
    "storeConfig": {
      "id": 1,
      "code": "default",
      "website_id": 1,
      "locale": "en_US",
      "base_currency_code": "USD",
      "default_display_currency_code": "USD",
      "timezone": "UTC",
      "weight_unit": "lbs",
      "base_url": "http://magento2/",
      "base_link_url": "http://magento2/default/",
      "base_static_url": "http://magento2/pub/static/version1590678667/",
      "base_media_url": "http://magento2/pub/media/",
      "secure_base_url": "https://magento2/",
      "secure_base_link_url": "https://magento2/default/",
      "secure_base_static_url": "https://magento2/pub/static/version1590678667/",
      "secure_base_media_url": "https://magento2/pub/media/",
      "store_name": null,
      "stores": [
        {
          "id": 1,
          "code": "default",
          "website_id": 1,
          "locale": "en_US",
          "base_currency_code": "USD",
          "default_display_currency_code": "USD",
          "timezone": "UTC",
          "weight_unit": "lbs",
          "base_url": "http://magento2/",
          "base_link_url": "http://magento2/default/",
          "base_static_url": "http://magento2/pub/static/version1590678667/",
          "base_media_url": "http://magento2/pub/media/",
          "secure_base_url": "https://magento2/",
          "secure_base_link_url": "https://magento2/default/",
          "secure_base_static_url": "https://magento2/pub/static/version1590678667/",
          "secure_base_media_url": "https://magento2/pub/media/",
          "store_name": null,
          "stores": null
        },
        {
          "id": 2,
          "code": "new_store_view",
          "website_id": 1,
          "locale": "en_US",
          "base_currency_code": "USD",
          "default_display_currency_code": "USD",
          "timezone": "UTC",
          "weight_unit": "lbs",
          "base_url": "http://magento2/",
          "base_link_url": "http://magento2/new_store_view/",
          "base_static_url": "http://magento2/pub/static/version1590678667/",
          "base_media_url": "http://magento2/pub/media/",
          "secure_base_url": "https://magento2/",
          "secure_base_link_url": "https://magento2/new_store_view/",
          "secure_base_static_url": "https://magento2/pub/static/version1590678667/",
          "secure_base_media_url": "https://magento2/pub/media/",
          "store_name": null,
          "stores": null
        }
      ]
    }
  }
}
```

swnsma reviewed (changes requested):
Hello @gallyamov!
Please check my comments.

gallyamov committed:
Merge branch '2.4-develop' of github.com:magento/magento2 into 28569_adding_stores_to_store_config

gallyamov committed:
magento/magento2#28569: Missing store codes in relation to a group and website

- Refactored code to use getList

gallyamov committed:
magento/magento2#28569: Missing store codes in relation to a group and website

- Added param type hint

gallyamov commented:
@magento run all tests

danielrenaud commented:
Hi @gallyamov 
I'm not sure this is the best approach from a schema perspective. I've opened up an architectural proposal with an alternative approach: https://github.com/magento/architecture/pull/390

danielrenaud commented:
@gallyamov Can you please implement this using the schema defined here: https://github.com/magento/architecture/pull/390

gallyamov committed:
Merge branch '2.4-develop' of github.com:magento/magento2 into 28569_adding_stores_to_store_config

gallyamov committed:
magento/magento2#28569: Multi-store: Missing store codes in relation to a group and website

- Implemented new availableStores query

- Code refactor

gallyamov commented:
@magento run all tests

danielrenaud commented on app/code/Magento/Store/Model/Service/StoreConfigManager.php:
```diff
@@ -71,12 +82,14 @@ public function getStoreConfigs(array $storeCodes = null)
+     * @return StoreConfigInterface
      */
-    protected function getStoreConfig($store)
+    public function getStoreConfig($store)
```
I don't think we need this change, and I would prefer to avoid it

danielrenaud replied:
I know we discussed this offline and thought it might not be a big deal.
However, now the class's api differs from it's interface, and updating the interface would be backwards incompatible.

cpartica replied:
you can cache values in getStoreConfigs, so it fetches data just once
then you can call `getStoreConfigs([$store->getCode()])` instead of calling getStoreConfig($store)
It's the same thing

danielrenaud commented on app/code/Magento/StoreGraphQl/Model/Resolver/Store/StoreConfigDataProvider.php:
```diff
@@ -55,24 +56,36 @@ public function __construct(
     {
-        $storeConfig = current($this->storeConfigManager->getStoreConfigs([$store->getCode()]));
+        $storesConfigData = [];
+        $storeConfigs = $this->storeConfigManager->getStoreConfigs();
```
We should only return available stores for the current website. This is returning all stores for all websites

danielrenaud commented on dev/tests/api-functional/testsuite/Magento/GraphQl/Store/StoreConfigResolverTest.php:
```diff
@@ -70,27 +76,85 @@ public function testGetStoreConfig()
+     * @magentoConfigFixture test_store store/information/name Test Store
+     * @throws Exception
+     */
+    public function testAvailableStoreConfigs(): void
```
I think this should be in a separate test file since it's testing a different query/resolver

gallyamov committed:
Merge branch '2.4-develop' of github.com:magento/magento2 into 28569_adding_stores_to_store_config

gallyamov committed:
magento/magento2#28569: Multi-store: Missing store codes in relation to a group and website

- Created separate test for available stores

Added website specific store output

gallyamov commented:
@magento run all tests

cpartica commented on app/code/Magento/Store/Model/Service/StoreConfigManager.php:
```diff
@@ -38,23 +47,25 @@ class StoreConfigManager implements \Magento\Store\Api\StoreConfigManagerInterfa
+     * @return StoreConfigInterface[]
      */
     public function getStoreConfigs(array $storeCodes = null)
     {
```
you can store StoreConfigs into a private local variable once fetched

danielrenaud commented on app/code/Magento/StoreGraphQl/Model/Resolver/Store/StoreConfigDataProvider.php:
```diff
@@ -55,24 +74,42 @@ public function __construct(
+    public function getAvailableStoreConfig(StoreInterface $store): array
     {
-        $storeConfig = current($this->storeConfigManager->getStoreConfigs([$store->getCode()]));
+        $storeIds = $this->storeWebsiteRelation->getStoreByWebsiteId($store->getWebsiteId());
```
Could we add a method to \Magento\Store\Model\ResourceModel\StoreWebsiteRelation that return the code instead of id (e.g. getStoreCodesByWebsiteId), to get the store codes.
then use StoreConfigManager::getStoreConfigs($code) to get the config info.
then we can avoid changing the protected method

danielrenaud commented on dev/tests/api-functional/testsuite/Magento/GraphQl/Store/StoreConfigResolverTest.php:
```diff
@@ -7,47 +7,55 @@
 
 /**
- * Test the GraphQL endpoint's StoreConfigs query
+ * Test the GraphQL endpoint's StoreConfigs and AvailableStores queries
```
Can you correct the test descriptions to be accurate

danielrenaud commented on app/code/Magento/StoreGraphQl/Model/Resolver/Store/StoreConfigDataProvider.php:
```diff
@@ -55,24 +74,42 @@ public function __construct(
      * @return array
      */
-    private function getBaseConfigData(StoreInterface $store) : array
+    public function getAvailableStoreConfig(StoreInterface $store): array
```
Maybe this should take just the website id as an argument. We don't need anything else from the store, and it makes more sense to me since we're getting available stores for a particular website

danielrenaud reviewed (changes requested):
please see review comments

gallyamov committed:
magento/magento2#28569: Multi-store: Missing store codes in relation to a group and website

- Code style and logic fixes

gallyamov commented:
@magento run all tests

swnsma commented on app/code/Magento/Store/Model/ResourceModel/StoreWebsiteRelation.php:
```diff
@@ -41,4 +41,29 @@ public function getStoreByWebsiteId($websiteId)
+     * @param bool $available
+     * @return array
+     */
+    public function getWebsiteStoreCodes($websiteId, $available = false): array
```
```suggestion
    public function getWebsiteStoreCodes(int $websiteId, bool $available = false): array
```

cpartica commented on app/code/Magento/StoreGraphQl/Model/Resolver/Store/StoreConfigDataProvider.php:
```diff
@@ -55,24 +65,38 @@ public function __construct(
      * @return array
      */
-    private function getBaseConfigData(StoreInterface $store) : array
+    public function getAvailableStoreConfig($websiteId): array
```
string strict type

cpartica commented on app/code/Magento/StoreGraphQl/Model/Resolver/Store/StoreConfigDataProvider.php:
```diff
@@ -83,14 +107,13 @@ private function getBaseConfigData(StoreInterface $store) : array
             'base_url' => $storeConfig->getBaseUrl(),
             'base_link_url' => $storeConfig->getBaseLinkUrl(),
-            'base_static_url' => $storeConfig->getSecureBaseStaticUrl(),
+            'base_static_url' => $storeConfig->getBaseStaticUrl(),
```
Please respond to this question:
what is the difference between between getSecureBaseStaticUrl and getBaseStaticUrl?
shouldn't this change along with magento setting to always use secure or not?
Does getBaseStaticUrl guarantee this? (same for the other urls)

cpartica commented on dev/tests/api-functional/testsuite/Magento/GraphQl/Store/AvailableStoreConfigTest.php:
```diff
@@ -0,0 +1,171 @@
+    public function testDefaultWebsiteAvailableStoreConfigs(): void
+    {
+        /** @var StoreConfigManagerInterface $storeConfigManager */
+        $storeConfigManager = $this->objectManager->get(StoreConfigManagerInterface::class);
```
move this to setup and private property

gallyamov committed:
Merge branch '2.4-develop' into 28569_adding_stores_to_store_config

gallyamov committed:
magento/magento2#28569: Multi-store: Missing store codes in relation to a group and website

- Fixed merge conflicts with #27952

gallyamov commented:
@magento run all tests

gallyamov committed:
magento/magento2#28569: Multi-store: Missing store codes in relation to a group and website

- Added strict types

gallyamov commented:
@magento run all tests

gallyamov committed:
magento/magento2#28569: Multi-store: Missing store codes in relation to a group and website

- Fixed static code checks and web api tests

gallyamov commented:
@magento run all tests

gallyamov committed:
magento/magento2#28569: Multi-store: Missing store codes in relation to a group and website

- Static code and test fixes

gallyamov commented:
@magento run all tests

swnsma committed:
Merge branch '2.4-develop' into 28569_adding_stores_to_store_config

gallyamov commented:
@magento run all tests

danielrenaud reviewed (dismissed):
Looks good.
SVC failure will need to be approved, but it looks like a reasonable change.

gallyamov committed:
Merge branch '2.4-develop' of github.com:magento/magento2 into 28569_adding_stores_to_store_config

gallyamov committed:
Merge branch '2.4-develop' into 28569_adding_stores_to_store_config

gallyamov committed:
magento/magento2#28569: Multi-store: Missing store codes in relation to a group and website

- Replaced interface updates

gallyamov committed:
Merge branch '28569_adding_stores_to_store_config' of github.com:magento/magento2 into 28569_adding_stores_to_store_config

gallyamov commented:
@magento run all tests

gallyamov committed:
magento/magento2#28569: Multi-store: Missing store codes in relation to a group and website

- Static code fix

gallyamov commented:
@magento run all tests

gallyamov committed:
magento/magento2#28569: Multi-store: Missing store codes in relation to a group and website

- Fixed store config API test

gallyamov commented:
@magento run all tests

swnsma commented:
@magento run all tests

swnsma commented on app/code/Magento/Store/Model/ResourceModel/StoreWebsiteRelation.php:
```diff
@@ -41,4 +43,29 @@ public function getStoreByWebsiteId($websiteId)
+     * @param bool $available
+     * @return array
+     */
+    public function getWebsiteStores(string $websiteId, bool $available = false): array
```
Why `$websiteId` has strict type `string`?
Generally, it represented via `int` strict types in the system.
My suggestion is to avoid ambiguous approach to do not 'surprise' anyone.
```suggestion
    public function getWebsiteStores(int $websiteId, bool $available = false): array
```

swnsma commented on app/code/Magento/StoreGraphQl/Model/Resolver/Store/StoreConfigDataProvider.php:
```diff
@@ -55,23 +65,42 @@ public function __construct(
      * @return array
      */
-    private function getBaseConfigData(StoreInterface $store) : array
+    public function getAvailableStoreConfig(string $websiteId): array
```
The same as already mentioned earlier.
```suggestion
    public function getAvailableStoreConfig(int $websiteId): array
```

cpartica commented on app/code/Magento/StoreGraphQl/Model/Resolver/Store/StoreConfigDataProvider.php:
```diff
@@ -55,23 +65,42 @@ public function __construct(
+    public function getAvailableStoreConfig(string $websiteId): array
     {
-        $storeConfig = current($this->storeConfigManager->getStoreConfigs([$store->getCode()]));
+        $websiteStores = $this->storeWebsiteRelation->getWebsiteStores($websiteId, true);
```
you need a (int) typecast, and plz do that int $websiteID in methods, it's the preferred way

cpartica commented:
@gallyamov, FYI "testDefaultWebsiteAvailableStoreConfigs" test is failing

gallyamov committed:
Merge branch '2.4-develop' of github.com:magento/magento2 into 28569_adding_stores_to_store_config

gallyamov committed:
magento/magento2#28569: Multi-store: Missing store codes in relation to a group and website

- Fixed strict types and tests

gallyamov commented:
@magento run all tests

magento-engcom-team commented:
Hi @swnsma, thank you for the review.
[ENGCOM-7810](https://jira.corp.magento.com/browse/ENGCOM-7810) has been created to process this Pull Request
